design-proposals/volume-migration/volume-migration.md

```diff
@@ -0,0 +1,297 @@
+
+## Repos
+
+  - [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+This proposal introduces a new CRD called `VolumeMigration`
+
+```golang
+type VolumeMigration struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec VolumeMigrationSpec `json:"spec"`
+
+	// +optional
+	Status *VolumeMigrationStatus `json:"status,omitempty"`
+}
+
+type VolumeMigrationList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	// +listType=atomic
+	Items []VolumeMigration `json:"items"`
+}
+
+// ReclaimPolicySourcePvc describes how the source PVC will be treated after the storage migration completes.
+// The policies follows the same behavior as the RetainPolicy for PVs: https://kubernetes.io/docs/concepts/storage/persistent-volumes/#reclaiming
+type ReclaimPolicySourcePvc string
+
+const (
+	DeleteReclaimPolicySourcePvc ReclaimPolicySourcePvc = "Delete"
+	RetainReclaimPolicySourcePvc ReclaimPolicySourcePvc = "Retain"
+)
+
+type MigratedVolume struct {
+	SourcePvc      string `json:"sourcePvc,omitempty" valid:"required"`
+	DestinationPvc string `json:"destinationPvc,omitempty" valid:"required"`
+	// ReclaimPolicySourcePvc describes how the source volumes will be
+	// treated after a successful migration
+	ReclaimPolicySourcePvc ReclaimPolicySourcePvc `json:"reclaimPolicySourcePvc,omitempty"`
+}
+
+type VolumeMigrationSpec struct {
+	// MigratedVolumes is a list of volumes to be migrated
+	MigratedVolume []MigratedVolume `json:"migratedVolume,omitempty"`
+}
+
+type VolumeMigrationState struct {
+	// MigratedVolumes is a list of volumes to be migrated
+	MigratedVolume []MigratedVolume `json:"migratedVolume,omitempty"`
+	// VirtualMachineMigrationName state of the virtual machine migration
+	// triggered by the storage migration
+	VirtualMachineMigrationName string `json:"virtualMachineMigrationName,omitempty"`
+	// VirtualMachineMigrationState state of the virtual machine migration
+	// triggered by the storage migration
+	VirtualMachineInstanceName string `json:"virtualMachineInstanceName,omitempty"`
+	Succeeded bool `json:"succeeded,omitempty"`
+	Failed bool `json:"failed,omitempty"`
+	// The time the migration action began
+	StartTimestamp *metav1.Time `json:"startTimestamp,omitempty"`
+	// The time the migration action ended
+	EndTimestamp *metav1.Time `json:"endTimestamp,omitempty"`
+}
+
+type RejectedMigratedVolume struct {
+	SourcePvc string `json:"sourcePvc,omitempty" valid:"required"`
+	Reason    string `json:"reason,omitempty" valid:"required"`
+}
+
+type VolumeMigrationStatus struct {
+    // VolumeMigrationStates describes the migration of the migrated volumes
+	VolumeMigrationStates []VolumeMigrationState   `json:"volumeMigrationStates,omitempty"`
+	// PendingVolumes are the migrated volumes that aren't associated to a VMI and cannot by migrated with this APIs
```

Comment:
typo `by` should be `be`
Thinking about it more, I thinking we should change the definition of `PendingVolumes` to volumes with one of the following properties:
- The target volume doesn't exist
- The migration hasn't started yet for whatever reason (maybe max concurrent reached, etc).
- The VM is not started, and thus cannot be migrated by the API

Reply:
Maybe also have the reason for the pending in the struct. Not sure if we should make it a seperate struct or add a field to MigratedVolume

Reply:
I don't know if it is necessary to extend the PendingVolumes

>     * The target volume doesn't exist
> 

If the target PVC doesn't exist then the target pod remains pending waiting for the PVC to be created. Then, the VMI migration stays in `Scheduling` state and eventually in `Failed` state. While the VMI migration is in `Scheduling` this is counted under the NotStarted.

I could add more information. However, the user can query the VMI migration to get all the information on a particular case.

>     * The migration hasn't started yet for whatever reason (maybe max concurrent reached, etc).
> 

Again, this is counted under NotStarted`.

>     * The VM is not started, and thus cannot be migrated by the API

In this case, there is no VMI and the migration doesn't take place and it is counted in `Pending`

Reply:
So what is the difference between pending and not started then? IIRC for pods pending encapsulates all the reasons I mentioned above. I am trying to reduce the number of lists if possible.

Reply:
Pending means that there are no VMI associated to the PVCs. Hence, either the user needs to start the VM or we need to use another method to perform the offline migration.
`NotStarted` means that there is a VM and a VirtualMachineInstanceMigration has been created but for some reason the migration hasn't started yet.